src/charm.py

```diff
@@ -0,0 +1,176 @@
+#!/usr/bin/env python3
+
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Charm for Synapse on kubernetes."""
+
+import logging
+import typing
+
+import ops
+from charms.nginx_ingress_integrator.v0.nginx_route import require_nginx_route
+from charms.traefik_k8s.v1.ingress import IngressPerAppRequirer
+from ops.charm import ActionEvent
+from ops.main import main
+
+import actions
+import synapse
+from charm_state import CharmConfigInvalidError, CharmState
+from constants import SYNAPSE_CONTAINER_NAME, SYNAPSE_NGINX_CONTAINER_NAME, SYNAPSE_NGINX_PORT
+from database_observer import DatabaseObserver
+from mjolnir import Mjolnir
+from observability import Observability
+from pebble import PebbleService, PebbleServiceError
+from saml_observer import SAMLObserver
+
+logger = logging.getLogger(__name__)
+
+
+class SynapseCharm(ops.CharmBase):
+    """Charm the service."""
+
+    def __init__(self, *args: typing.Any) -> None:
+        """Construct.
+
+        Args:
+            args: class arguments.
+        """
+        super().__init__(*args)
+        self.database = DatabaseObserver(self)
+        self.saml = SAMLObserver(self)
+        try:
+            self._charm_state = CharmState.from_charm(charm=self)
+        except CharmConfigInvalidError as exc:
+            self.model.unit.status = ops.BlockedStatus(exc.msg)
+            return
+        self.pebble_service = PebbleService(charm_state=self._charm_state)
+        # service-hostname is a required field so we're hardcoding to the same
+        # value as service-name. service-hostname should be set via Nginx
+        # Ingress Integrator charm config.
+        require_nginx_route(
+            charm=self,
+            service_hostname=self.app.name,
+            service_name=self.app.name,
+            service_port=SYNAPSE_NGINX_PORT,
+        )
+        self._ingress = IngressPerAppRequirer(
+            self,
+            port=SYNAPSE_NGINX_PORT,
+            # We're forced to use the app's service endpoint
+            # as the ingress per app interface currently always routes to the leader.
+            # https://github.com/canonical/traefik-k8s-operator/issues/159
+            host=f"{self.app.name}-endpoints.{self.model.name}.svc.cluster.local",
+            strip_prefix=True,
+        )
+        self._observability = Observability(self)
+        # Mjolnir is a moderation tool for Matrix.
+        # See https://github.com/matrix-org/mjolnir/ for more details about it.
+        if self._charm_state.enable_mjolnir:
+            self._mjolnir = Mjolnir(self, charm_state=self._charm_state)
```

Comment:
This was changed based on this discussion https://github.com/canonical/synapse-operator/pull/30#discussion_r1324711002
I believe this will be changed after the refactoring regarding splitting in Service module/Observer (ISD-1095)

Reply:
Related: ISD-1095